src/osdc/Objecter.cc

```diff
@@ -3058,8 +3058,10 @@ MOSDOp *Objecter::_prepare_osd_op(Op *op)
 
   int flags = op->target.flags;
   flags |= CEPH_OSD_FLAG_KNOWN_REDIR;
-  if (op->onfinish)
-    flags |= CEPH_OSD_FLAG_ONDISK;
+
+  // Nothing checks this any longer, but needed for compatibility with
+  // pre-luminous osds
+  flags |= CEPH_OSD_FLAG_ONDISK;
```

Comment:
1) there is now only one kind of reply (there used to be 'ack' and 'ondisk' replies)
2) this changes makes it so we always request a reply for an operation.  this particular bug was that we didn't request an ack, which meant that the client would never close it out and we would resend the request later when the pg moved.
3) the other patches simplify cache flushing, which was originally more complicated because it was working around the bug (before we understood what was going on).